svm/src/account_loader.rs

```diff
@@ -2227,4 +2371,271 @@ mod tests {
+            test_data_size(ixns, bpf_loader_size + fee_payer_size);
+
+            // native loader counted if instruction
+            let ixns = vec![Instruction::new_with_bytes(
+                bpf_loader::id(),
+                &[],
+                vec![account_meta(native_loader::id(), false)],
+            )];
+            test_data_size(ixns, bpf_loader_size + native_loader_size + fee_payer_size);
+
+            // loader counted twice if included in instruction
+            let ixns = vec![Instruction::new_with_bytes(
+                program1,
+                &[],
+                vec![account_meta(bpf_loader::id(), false)],
+            )];
+            test_data_size(ixns, program1_size + bpf_loader_size * 2 + fee_payer_size);
+
+            // cover that case with multiple loaders to be sure
+            let ixns = vec![
+                Instruction::new_with_bytes(
+                    program1,
+                    &[],
+                    vec![
+                        account_meta(bpf_loader::id(), false),
+                        account_meta(bpf_loader_upgradeable::id(), false),
+                    ],
+                ),
+                Instruction::new_with_bytes(program2, &[], vec![account_meta(account1, false)]),
+                Instruction::new_with_bytes(
+                    bpf_loader_upgradeable::id(),
+                    &[],
+                    vec![account_meta(account1, false)],
+                ),
+            ];
+            test_data_size(
+                ixns,
+                account1_size
+                    + program1_size
+                    + program2_size
+                    + bpf_loader_size * 2
+                    + upgradeable_loader_size * 2
+                    + fee_payer_size,
+            );
+
+            // loader counted twice even if included first
+            let ixns = vec![
+                Instruction::new_with_bytes(bpf_loader::id(), &[], vec![]),
+                Instruction::new_with_bytes(program1, &[], vec![]),
+            ];
+            test_data_size(ixns, program1_size + bpf_loader_size * 2 + fee_payer_size);
+
+            // fee-payer counted once
+            let ixns = vec![Instruction::new_with_bytes(
+                program1,
+                &[],
+                vec![account_meta(fee_payer, false)],
+            )];
+            test_data_size(ixns, program1_size + bpf_loader_size + fee_payer_size);
+
+            // edge cases involving program cache
+            let mut program_cache = ProgramCacheForTxBatch::default();
+
+            let program2_entry = ProgramCacheEntry {
+                account_size: (program2_size + programdata2_size) as usize,
+                account_owner: ProgramCacheEntryOwner::LoaderV3,
+                ..ProgramCacheEntry::default()
+            };
+            program_cache.replenish(program2, Arc::new(program2_entry));
+
+            let programdata2_entry = ProgramCacheEntry {
```

Comment:
the programdata has a separate entry in the cache? is this what actually happens in the cache?

Reply:
it depends... it shows up in the cache as a buffer account, stays there after you deploy the program, but then if its later evicted it doesnt reenter the cache again. ill remove this tho, it shouldnt be relevant to this test since the sizes dont differ